sled-agent/src/opte/illumos/firewall_rules.rs

```diff
@@ -24,38 +24,38 @@ use oxide_vpc::api::ProtoFilter;
 use oxide_vpc::api::Protocol;
 
 trait FromVpcFirewallRule {
-    fn action(self: &Self) -> Action;
```

Comment:
Why doesn't `rustc` catch this? Shouldn't the ref-ness of the variable called `self` exactly match that of its type? IOW, I'd expect both `self: Self` and `&self: &Self` to be valid, but I somehow managed to fat-finger the combination I think makes the least sense.

Reply:
I don't think `&self: &Self` means what you think it does. Stepping away from `self`/`Self` entirely for a second for simplicity, this function takes an `i32` ref, and inside the body of the function the type of `x` is `&i32`:

```rust
fn foo(x: &i32) {
    println!("{}", std::any::type_name_of_val(&x));
}
```

This function also takes an `i32` ref, but inside the body of the function the type of `x` is `i32`:

```rust
fn bar(&x: &i32) {
    println!("{}", std::any::type_name_of_val(&x));
}
```

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=60a200507616f8cc2710db49e2a3bd2c)

A function arg written as `&x: &T` is a [pattern match](https://doc.rust-lang.org/book/ch18-01-all-the-places-for-patterns.html#function-parameters), just like you can write `for &x in &some_vec { .. }` to "strip off" a layer of references (which only works if the `T` implements `Copy`!). I didn't realize `self: T` worked for anything other than the special pointer types (`Arc` in particular is the only one I've used), but since it does, I think it follows the same rules:

```
self: &Self  // same as taking just `&self`
self: Self   // same as taking just `self`
&self: &Self // takes `&Self`, but requires `Self: Copy`, and in the body, `self` has type `Self`
```

Reply:
Ah, ok, that makes sense. Thanks for the explanation!